### PR TITLE
Feature/device commands and ack messages iotsol 71 iotsol 55

### PR DESCRIPTION
**Pull request by sw-tt-kirtanvaghela**

### CODE REVIEW

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift [hunk 1]
```diff
@@ -209,18 +227,19 @@ class IoTConnectManager {
...
-//            }else{
-//                print("prevsendData is nil")
+            //check for data frequency
+            if let time = prevSendDataTime{
+                let diff = Int(Date().timeIntervalSince(time))
+                print("df \(df) diff\(diff) \(time) \(Date())")
+                if diff >= df{
+                    validateData(data: data)
+                }else{
+                    print("Diff is lt")
```
Comment (sw-ot-deveshmevada) [inline IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift:237]: What happens if it goes in else case?

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift [hunk 2]
```diff
@@ -300,6 +319,27 @@ class IoTConnectManager {
         }
     }
     
+    //send ack command for ota,module and device command
+    func sendAckCmd(ackGuid:String,status:String, msg:String = "",childId:String = "",type:Int){
+        if dictSyncResponse.count > 0{
+            let dictToSend =  [
+                "dt":objCommon.now(),
```
Comment (sw-ot-deveshmevada) [inline IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift:326]: Please use cod able or use constants for keys

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift [hunk 3]
```diff
@@ -368,10 +408,14 @@ class IoTConnectManager {
                 if dictD?.isEmpty == true{
                     if let valDict = value as? [String:Any]{
                         for (valDictKey,valDictValue) in valDict{
-                            dictD?.append(anotherDict: [    "\(key)":[valDictKey:["\(valDictValue)","\(valDictValue)","\(valDictValue)","\(valDictValue)","1.0","\(valDictValue)"]]])
+//<<<<<<< Updated upstream
```
Comment (sw-ot-deveshmevada) [inline IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift:411]: remove unwanted code

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift [hunk 4]
```diff
@@ -427,10 +471,10 @@ class IoTConnectManager {
                                 // arrData.append([key:[
                                 //                valDictKey:[valDictValue]]])
                                 dictD?.append(anotherDict: [
-                                    "\(key)":[valDictKey:["\(valDictValue)","\(valDictValue)","\(valDictValue)","\(valDictValue)","1.0","\(valDictValue)"]]])
+                                    "\(key)":[valDictKey:["\(valDictValue)","\(valDictValue)","\(valDictValue)","\(valDictValue)","1","\(valDictValue)"]]])
```
Comment (sw-ot-deveshmevada) [inline IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift:474]: add "1" to constant

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift [hunk 5]
```diff
@@ -726,24 +771,61 @@ class IoTConnectManager {
...
     }
     
+    //called on data frequncy change
+    func onFrequencyChangeCommand(dfValue:Int){
+        var metaInfo = self.dictSyncResponse["meta"] as? [String:Any]
+        metaInfo?["df"] = dfValue
+//        print("metaInfo Before \(self.dictSyncResponse)")
+        self.dictSyncResponse["meta"] = metaInfo
+//        print("metaInfo after \(self.dictSyncResponse)")
+        df = metaInfo?["df"] as? Int ?? 0
```
Comment (sw-ot-deveshmevada) [inline IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift:801]: use constant for keys

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift [hunk 6]
```diff
@@ -726,24 +771,61 @@ class IoTConnectManager {
...
+        print("df changed val \(df)")
+    }
+    
+    //Publish topic on MQTT to create child device
+    public func createChildDevice(deviceId:String, deviceTag:String, displayName:String){
+        let metaInfo = self.dictSyncResponse["meta"] as? [String:Any]
+        let gtw = metaInfo?["gtw"] as? [String:Any]
+        let g = gtw?["g"] as? String
+        
+        objMQTTClient.publishTopicOnMQTT(withData: ["mt":CommandType.CREATE_DEVICE.rawValue,
```
Comment (sw-ot-deveshmevada) [inline IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift:811]: Same comment here

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectService.swift [hunk 7]
```diff
@@ -374,6 +406,9 @@ extension IoTConnectManager {
                             if dataDevice[keyPath:"d.rc"] as! Int == DeviceSync.Response.OK {
                                 var dictToUpdate = self.dictSyncResponse
                                 if strKey == CommandType.ATTRIBUTE_INFO_UPDATE.rawValue {
+//=======
```
Comment (sw-ot-deveshmevada) [inline IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectService.swift:409]: remove unwanted comment

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectService.swift [hunk 8]
```diff
@@ -1913,5 +1949,16 @@ extension IoTConnectManager {
                 print("Error parsing EdgeRule Response")
             }
         }
-    } 
+//<<<<<<< Updated upstream
```
Comment (sw-ot-deveshmevada) [inline IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectService.swift:1952]: remove unwanted comments

**Review by sw-ot-deveshmevada — CHANGES_REQUESTED** [review]

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift (hunk 1 above)
Comment (sw-tt-kirtanvaghela) [reply IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift:237]: If difference between time is less then df value from portal then nothing will happen means data is not going to be sent on MQTT.

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift (hunk 2 above)
Comment (sw-tt-kirtanvaghela) [reply IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift]: @sw-ot-deveshmevada Changes are done in new push

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift (hunk 3 above)
Comment (sw-tt-kirtanvaghela) [reply IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift]: @sw-ot-deveshmevada Changes are done in new push

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift (hunk 4 above)
Comment (sw-tt-kirtanvaghela) [reply IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift]: @sw-ot-deveshmevada Changes are done in new push

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift (hunk 6 above)
Comment (sw-tt-kirtanvaghela) [reply IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectManager.swift:633]: @sw-ot-deveshmevada Changes are done in new push

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectService.swift (hunk 7 above)
Comment (sw-tt-kirtanvaghela) [reply IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectService.swift]: @sw-ot-deveshmevada Changes are done in new push

File: IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectService.swift (hunk 8 above)
Comment (sw-tt-kirtanvaghela) [reply IoTConnect_2.0/IoTConnect_2.0/IoTConnectManager/IoTConnectService.swift]: @sw-ot-deveshmevada Changes are done in new push

**Pull request merged**